plugins/hangman/hangman.7
.\"
.\" Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd @IRCCD_MAN_DATE@
.Dt IRCCD-PLUGIN-HANGMAN 7
.Os
.\" NAME
.Sh NAME
.Nm hangman
.Nd irccd hangman plugin
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
plugin lets you play the hangman game on IRC.
.\" INSTALLATION
.Sh INSTALLATION
The plugin
.Nm
is distributed with irccd. To enable it add it into your
.Xr irccd.conf 5
configuration file.
.Pp
.Bd -literal -offset indent
plugin "hangman"
.Ed
.\" USAGE
.Sh USAGE
The
.Nm
plugin starts when a user execute its special command with no arguments.
.Bd -literal -offset indent
markand: !hangman
irccd: markand, the game is started, the word to find is: _ _ _ _ _ _ _ _ _ _ _
.Ed
.Pp
If a game is already running, the same command shows the current word.
.Bd -literal -offset indent
markand: !hangman
irccd: markand, the game is already running and the word is: s _ _,
.Ed
.Pp
There are two ways for proposing a response to the game. You can either just ask
for a letter or for a whole word.
.Ss By letter
When asking a letter, the message must be one letter.
.Bd -literal -offset Ds
markand: c
irccd: markand, nice! the word is now c _ _ _ _
jean: k
irccd: jean, there is no 'k'.
.Ed
.Ss By full word
When asking by a word, just put one word as command argument. Note that in
contrast to the letter proposal we have used command invocation rather than
normal message to avoid possible propositions by mistake.
.Bd -literal -offset Ds
markand: !hangman couch
irccd: markand, this is not the word.
jean: !hangman candy
irccd: jean, congratulations, the word is candy.
.Ed
.\" CONFIGURATION
.Sh CONFIGURATION
The plugin
.Nm
supports the following configuration properties:
.Bl -tag -width "collaborative (bool)"
.It Va file No (string)
The path to the database file.
.Pp
The file must contain one word per line, it must be saved as UTF-8 and words
must consist of UTF-8 characters, any other entry will be ignored.
.Pp
Example:
.Bd -literal
sky
irccd
FreeBSD
door
cat
.Ed
.It Va collaborative No (bool)
Set to true to enable collaborative mode, a player can't propose two
consecutives proposals (Optional, default: true).
.El
.\" TEMPLATES
.Sh TEMPLATES
The plugin
.Nm
supports the following template properties:
.Bl -tag -width wrong-letter
.It Va asked
When a letter has been already asked but present in the word. Keywords:
.Em letter .
.It Va dead
When the man was hung. Keywords:
.Em word .
.It Va found
When a correct letter has been placed. Keywords:
.Em word .
.It Va running
When a game is requested but it's already running. Keywords:
.Em word .
.It Va start
When the game starts. Keywords:
.Em word .
.It Va win
When the game succeeded. Keywords:
.Em word .
.It Va wrong-word
When a word proposal is wrong. Keywords:
.Em word .
.It Va wrong-letter
When a letter proposal is wrong. Keywords:
.Em letter .
.El
.Pp
All templates support the following standard keywords:
.Em channel , command , nickname , origin , plugin , server .
.\" EXAMPLES
.Sh EXAMPLES
Example of configuration file:
.Bd -literal
plugin "hangman" {
  config {
    "file" "/var/srv/db/words.txt";
  }

  templates {
    "win"          "nice job, the word was #{word}!";
    "wrong-letter" "please try again, there is no #{letter}";
  }
}
.Ed
.\" SEE ALSO
.Sh SEE ALSO
.Xr irccd 1 ,
.Xr irccd-templates 7


CHANGES.md
IRC Client Daemon CHANGES
=========================

irccd 5.0.0 ????-??-??
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight of this release is the integration of libev as the main loop
and coroutines, making the code simpler and more expressive.

Build system
------------

The build system has been changed from CMake to GNU Make, this is because CMake
makes simple things complicated.

C API
-----

- Most of the string based types are dynamically allocated rather than creating
  dozens of arbitrary fixed size limits.
- Functions for allocated types are now coupled as `foo_new` and `foo_free` or
  `foo_init` and `foo_finish` otherwise.
- Types that require string list are now implemented using NULL sentinel based
  lists.

irccd
-----

- The auto-reconnect feature is automatically enabled on a server and is no
  longer an option. As an IRC bot, it make sense that a server should be
  reconnected permanently.
- All deferred functions have been removed.
- The server connection procedure is now implemented using coroutines for a
  more readable code logic.

misc
----

The C and Javascript APIs are now documented using doxygen rather than manual
pages.

irccd 4.1.1 2026-01-07
======================

- Fix minor issue in Irccd.Http API when `curl_multi_fdset` can return `-1` file
  descriptor.

irccd 4.1.0 2025-04-23
======================

- Add `irc_bot_pollable_add` to insert a custom interface into the main irccd
  loop.
- A new option `IRCCD_WITH_HTTP` has been added to accommodate the new curl
  dependency for `Irccd.Http` API.
- The plugin `history` manages two different timestamps for message and
  visibility. This helps showing the last message date of someone who is present
  on a channel but idle.

javascript API
--------------

- Brand new asynchronous `Irccd.Http` API.

plugins
-------

- links: rewritten in Javascript using the brand new `Irccd.Http` API.

irccd 4.0.4 2023-08-03
======================

- Improve tests running on a slow machines.
- Fix a warning on a invalid comparison.

irccd 4.0.3 2023-02-08
======================

- Fix build on NetBSD 9.3.

irccd 4.0.2 2023-01-18
======================

- Fix crash because of undefined variable.

irccd 4.0.1 2022-02-07
======================

- Fix disconnection with multiple servers.
- Fix Javascript `Irccd.Server` constructor function.
- Substitute the irccd version in Javascript plugins.
- Fix `irccdctl plugin-reload` command without argument.

irccd 4.0.0 2022-02-03
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight for this release is the rewrite from C++ to C. The only
runtime dependency required is OpenSSL (if built with SSL support).

irccd
-----

- Irccd keeps track of nicknames in channels by capturing join/part/kick and
  mode changes. It is now more convenient from the plugins to quickly inspect if
  someone is present on a channel.
- It is now possible to change uid/gid of the transport socket file. The file is
  also created with permissions 664.
- A new `paths` command has been added to show default paths.

irccdctl
--------

- Commands `plugin-reload` and `plugin-unload` can be invoked without arguments.
- New `plugin-template` and `plugin-path` command which are synonyms of
  `plugin-config` but for templates and paths respectively.

plugins
-------

- tictactoe: now has a timeout in case of inactivity.

misc
----

- Split irccd-api manual page into individual irccd-api-<module> for a better
  readability.
- New `irccd.conf` and `irccdctl.conf` syntax.

network API
-----------

- Network protocol uses plain text again.
- Transport uses clear UNIX sockets only without passwords.

javascript API
--------------

- Brand new Irccd.Rule API to inspect and manage rules.
- Brand new Irccd.Hook API to inspect and manage hooks.

irccd 3.1.1 2021-01-04
======================

- Synchronize `ping-timeout` option in `[server]` to 1800 seconds by default.
- Enable `auto-reconnect` option in `[server]` by default as specified in the
  manual page.

irccd 3.1.0 2020-07-03
======================

- Added a new hook system. Hooks consist of an alternative approach to plugins
  to extend irccd in any language.

irccd 3.0.3 2019-10-06
======================

- Fix errors in irccdctl.conf example file.
- Add example of password in irccdctl.conf and irccd.conf.

irccd 3.0.2 2019-09-22
======================

- Added *IRCCD_WITH_JS* CMake variable in irccd package.
- Fixed trailing CTCP escape code.
- Fixed invalid parameters in topic event.
- Fixed invalid documentation of Irccd.System.usleep function.
- Fixed invalid system configuration directory.

irccd 3.0.1 2019-09-01
======================

- Fixed an invalid template escape sequence.
- Updated the default configuration files.
- Fix RPATH handling for private libraries like Duktape.

irccd 3.0.0 2019-08-15
======================

This is a major release. See MIGRATING.md file for more information.

irccd
-----

- New sections `[paths]` and `[paths.plugin]` have been added to control
  standard paths for both irccd and plugins.
- Irccd no longer supports uid, gid, pid and daemon features.
- Sections `[identity]` and `[server]` have been merged.
- Local transports support SSL.
- The origin in rule is now first class value.
- New option `ipv4` in `[transport]`.
- New option `ipv4` in `[server]`.
- Section `[format]` is renamed to `[templates]`.
- New commands are available as irccd arguments `info` and `version`.

irccdctl
--------

- New option `ipv4` in `[connect]`.
- New option `-o` in `rule-add`.
- New option `-o` and `-O` in `rule-edit`.

irccd-test
----------

- A brand new `irccd-test` program has been added to tests plugins on the
  command line.

cmake
-----

- CMake no longer create a fake installation directory while building.
- All targets are placed into the `bin` directory while building.

network API
-----------


plugins/hangman/hangman.js
/*
 * hangman.js -- hangman game for IRC
 *
 * Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

// Plugin information.
info = {
	author: "David Demelier <markand@malikania.fr>",
	license: "ISC",
	summary: "A hangman game for IRC",
	version: "@irccd_VERSION@"
};

// Modules.
var Logger = Irccd.Logger;
var File = Irccd.File;
var Plugin = Irccd.Plugin;
var Server = Irccd.Server;
var Unicode = Irccd.Unicode
var Util = Irccd.Util;

// Default options.
Plugin.config["collaborative"] = "true";

// Formats.
Plugin.templates = {
	"asked":        "#{nickname}, '#{letter}' was already asked.",
	"dead":         "#{nickname}, fail the word was: #{word}.",
	"found":        "#{nickname}, nice! the word is now #{word}",
	"running":      "#{nickname}, the game is already running and the word is: #{word}",
	"start":        "#{nickname}, the game is started, the word to find is: #{word}",
	"win":          "#{nickname}, congratulations, the word is #{word}.",
	"wrong-word":   "#{nickname}, this is not the word.",
	"wrong-player": "#{nickname}, please wait until someone else proposes.",
	"wrong-letter": "#{nickname}, there is no '#{letter}'."
};

function Hangman(server, channel)
{
	this.server = server;
	this.channel = channel;
	this.tries = 10;
	this.select();
}

/**
 * Map of games.
 */
Hangman.map = {};

/**
 * List of words.
 */
Hangman.words = {
	all: [],        //!< All words,
	registry: {}    //!< Words list per server/channel.
};

/**
 * Search for an existing game.
 *
 * @param server the server object
 * @param channel the channel name
 * @return the hangman instance or undefined if no one exists
 */
Hangman.find = function (server, channel)
{
	return Hangman.map[server.toString() + '@' + channel];
}

/**
 * Create a new game, store it in the map and return it.
 *
 * @param server the server object
 * @param channel the channel name
 * @return the hangman object
 */
Hangman.create = function (server, channel)
{
	return Hangman.map[server.toString() + "@" + channel] = new Hangman(server, channel);
}

/**
 * Remove the specified game from the map.
 *
 * @param game the game to remove
 */
Hangman.remove = function (game)
{
	delete Hangman.map[game.server + "@" + game.channel];
}

/**
 * Check if the text is a valid word.
 *
 * @param word the word to check
 * @return true if a word
 */
Hangman.isWord = function (word)
{
	if (word.length === 0)
		return false;

	for (var i = 0; i < word.length; ++i)
		if (!Unicode.isLetter(word.charCodeAt(i)))
			return false;

	return true;
}

/**
 * Load all words.
 */
Hangman.loadWords = function ()
{
	var path;

	// User specified file?
	if (Plugin.config["file"])
		path = Plugin.config["file"];
	else
		path = Plugin.paths.config + "/words.conf";

	try {
		Logger.info("loading words from " + path);

		var file = new File(path, "r");
		var line;

		Hangman.words.all = [];

		while ((line = file.readline()) !== undefined)
			if (Hangman.isWord(line))
				Hangman.words.all.push(line);
	} catch (e) {
		throw new Error("could not open '" + path + "'");
	}

	if (Hangman.words.all.length === 0)
		throw new Error("empty word database");

	Logger.info("number of words in database: " + Hangman.words.all.length);
}

/**
 * Select the next word for the game.
 */
Hangman.prototype.select = function ()
{
	var id = this.server.toString() + "@" + this.channel;

	// Reload the words if empty.
	if (!Hangman.words.registry[id] || Hangman.words.registry[id].length === 0)
		Hangman.words.registry[id] = Hangman.words.all.slice(0);

	var i = Math.floor(Math.random() * Hangman.words.registry[id].length);

	this.word = Hangman.words.registry[id][i];

	// Erase words from the registry.
	Hangman.words.registry[id].splice(i, 1);

	// Fill table.
	this.table = {};

	for (var j = 0; j < this.word.length; ++j)
		this.table[this.word.charCodeAt(j)] = false;
}

/**
 * Format the word with underscore and letters.
 *
 * @return the secret
 */
Hangman.prototype.formatWord = function ()
{
	var str = "";

	for (var i = 0; i < this.word.length; ++i) {
		var ch = this.word.charCodeAt(i);

		if (!this.table[ch])
			str += "_";
		else
			str += String.fromCharCode(ch);

		if (i + 1 < this.word.length)
			str += " ";
	}

	return str;
}

/**
 * Propose a word or a letter.
 *
 * @param ch the code point or the unique word
 * @param nickname the user trying
 * @return the status of the game
 */
Hangman.prototype.propose = function (ch, nickname)
{
	var status = "found";

	// Check for collaborative mode.
	if (Plugin.config["collaborative"] === "true" && !this.query) {
		if (this.last !== undefined && this.last === nickname)
			return "wrong-player";

		this.last = nickname;
	}

	if (typeof(ch) == "number") {
		if (this.table[ch] === undefined) {
			this.tries -= 1;
			status = "wrong-letter";
		} else {
			if (this.table[ch]) {
				this.tries -= 1;
				status = "asked";
			} else
				this.table[ch] = true;
		}
	} else {
		if (this.word != ch) {
			this.tries -= 1;
			status = "wrong-word";
		} else
			status = "win";
	}

	// Check if dead.
	if (this.tries <= 0)
		status = "dead";

	// Check if win.
	var win = true;

	for (var i = 0; i < this.word.length; ++i) {
		if (!this.table[this.word.charCodeAt(i)]) {
			win = false;
			break;
		}
	}

	if (win)
		status = "win";

	return status;
}

function onLoad()
{
	Hangman.loadWords();
}

onReload = onLoad;

function propose(server, channel, origin, game, proposition)
{
	var kw = {
		channel: channel,
		command: server.info().prefix + Plugin.info().name,
		nickname: Util.splituser(origin),
		origin: origin,
		plugin: Plugin.info().name,
		server: server.toString()
	};

	var st = game.propose(proposition, kw.nickname);

	switch (st) {
	case "found":
		kw.word = game.formatWord();
		server.message(channel, Util.format(Plugin.templates["found"], kw));
		break;
	case "wrong-letter":
	case "wrong-player":
	case "wrong-word":
		kw.word = proposition;
	case "asked":
		kw.letter = String.fromCharCode(proposition);
		server.message(channel, Util.format(Plugin.templates[st], kw));
		break;
	case "dead":
	case "win":
		kw.word = game.word;
		server.message(channel, Util.format(Plugin.templates[st], kw));

		// Remove the game.
		Hangman.remove(game);
		break;
	default:
		break;
	}
}

function onCommand(server, origin, channel, message)
{
	var isquery = server.isSelf(channel);

	if (isquery)
		channel = origin;
	else
		channel = channel.toLowerCase();

	var game = Hangman.find(server, channel);
	var kw = {
		channel: channel,
		command: server.info().prefix + Plugin.info().name,
		nickname: Util.splituser(origin),
		origin: origin,
		plugin: Plugin.info().name,
		server: server.toString()
	};

	if (game) {
		var list = message.split(" \t");

		if (list.length === 0 || String(list[0]).length === 0) {
			kw.word = game.formatWord();
			server.message(channel, Util.format(Plugin.templates["running"], kw));
		} else {
			var word = String(list[0]);

			if (Hangman.isWord(word))
				propose(server, channel, origin, game, word);
		}
	} else {
		game = Hangman.create(server, channel);
		game.query = isquery;
		kw.word = game.formatWord();
		server.message(channel, Util.format(Plugin.templates["start"], kw));
	}

	return game;
}

function onMessage(server, origin, channel, message)
{
	if (server.isSelf(channel))
		channel = origin;
	else
		channel = channel.toLowerCase();

	var game = Hangman.find(server, channel);

	if (!game)
		return;

	if (message.length === 1 && Unicode.isLetter(message.charCodeAt(0)))
		propose(server, channel, origin, game, message.charCodeAt(0));
}
